src/Microsoft.Health.Fhir.SqlServer/Features/Storage/TvpRowGeneration/BulkStringSearchParameterV3RowGenerator.cs

```diff
@@ -0,0 +1,50 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Security.Cryptography;
+using System.Text;
+using Microsoft.Health.Fhir.Core.Features.Search.SearchValues;
+using Microsoft.Health.Fhir.SqlServer.Features.Schema.Model;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Storage.TvpRowGeneration
+{
+    internal class BulkStringSearchParameterV3RowGenerator : BulkSearchParameterRowGenerator<StringSearchValue, BulkStringSearchParamTableTypeV3Row>
+    {
+        private readonly int _indexedTextMaxLength = (int)VLatest.StringSearchParam.Text.Metadata.MaxLength;
+
+        public BulkStringSearchParameterV3RowGenerator(SqlServerFhirModel model, SearchParameterToSearchValueTypeMap searchParameterTypeMap)
+            : base(model, searchParameterTypeMap)
+        {
+        }
+
+        internal override bool TryGenerateRow(int offset, short searchParamId, StringSearchValue searchValue, out BulkStringSearchParamTableTypeV3Row row)
+        {
+            string indexedPrefix;
+            string overflow;
+            if (searchValue.String.Length > _indexedTextMaxLength)
+            {
+                // TODO: this truncation can break apart grapheme clusters.
+                indexedPrefix = searchValue.String.Substring(0, _indexedTextMaxLength);
+                overflow = searchValue.String;
+            }
+            else
+            {
+                indexedPrefix = searchValue.String;
+                overflow = null;
+            }
+
+            string computedHash;
+            using (SHA256 sha256Hash = SHA256.Create())
```

Comment:
>SHA256.Create() [](http://example.com/codeflow?start=39&length=15)

I wonder, how heavy it is to create this object on each row.

Ideally we should create it once, but it's probably mean we need to make generator disposable, and utilize it, which is a lot of work...

Reply:
Thanks, updated to implement IDisposable

Reply:
Unfortunately it's a bit trickier than just adding `IDisposable` to a class. Some one need to call it.

For this particular case, in this particular time, it will work fine, since we have only one instance of it, because we inject it with DI as singleton
see `AddSqlServerTableRowParameterGenerators`
so it would be freed up during app closure.